src/wamr/wamr.cc

```diff
@@ -0,0 +1,767 @@
+      const auto section_data_start = pos;
+      const auto section_name_len = parseVarint(pos, end);
+      if (section_name_len == static_cast<uint32_t>(-1) || pos + section_name_len > end) {
+        return false;
+      }
+      auto section_name = std::string_view(pos, section_name_len);
+      if (section_name.find("precompiled_") != std::string::npos) {
+        // If this is the first "precompiled_" section, then save everything
+        // before it, otherwise skip it.
+        if (stripped.empty()) {
+          const byte_t *start = source_.get()->data;
+          stripped.insert(stripped.end(), start, section_start);
+        }
+      }
+      pos = section_data_start + section_len;
+    } else {
+      pos += section_len;
+      // Save this section if we already saw a custom "precompiled_" section.
+      if (!stripped.empty()) {
+        stripped.insert(stripped.end(), section_start, pos /* section end */);
+      }
+    }
+  }
+
+  if (!stripped.empty()) {
+    wasm_byte_vec_new_uninitialized(out->get(), stripped.size());
+    ::memcpy(out->get()->data, stripped.data(), stripped.size());
+    return true;
+  }
+  return false;
+}
+
+static bool equalValTypes(const wasm_valtype_vec_t *left, const wasm_valtype_vec_t *right) {
+  if (left->size != right->size) {
+    return false;
+  }
+
+  for (size_t i = 0; i < left->size; i++) {
+    if (wasm_valtype_kind(left->data[i]) != wasm_valtype_kind(right->data[i])) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+static std::string printValue(const wasm_val_t &value) {
+  switch (value.kind) {
+  case WASM_I32:
+    return std::to_string(value.of.i32);
+  case WASM_I64:
+    return std::to_string(value.of.i64);
+  case WASM_F32:
+    return std::to_string(value.of.f32);
+  case WASM_F64:
+    return std::to_string(value.of.f64);
+  default:
+    return "unknown";
+  }
+}
+
+static std::string printValues(const wasm_val_t values[], size_t size) {
+  if (size == 0) {
+    return "";
+  }
+
+  std::string s;
+  for (size_t i = 0; i < size; i++) {
+    if (i) {
+      s.append(", ");
+    }
+    s.append(printValue(values[i]));
+  }
+  return s;
+}
+
+static const char *printValKind(wasm_valkind_t kind) {
+  switch (kind) {
+  case WASM_I32:
+    return "i32";
+  case WASM_I64:
+    return "i64";
+  case WASM_F32:
+    return "f32";
+  case WASM_F64:
+    return "f64";
+  case WASM_ANYREF:
+    return "anyref";
+  case WASM_FUNCREF:
+    return "funcref";
+  default:
+    return "unknown";
+  }
+}
+
+static std::string printValTypes(const wasm_valtype_vec_t *types) {
+  if (types->size == 0) {
+    return "void";
+  }
+
+  std::string s;
+  s.reserve(types->size * 8 /* max size + " " */ - 1);
+  for (size_t i = 0; i < types->size; i++) {
+    if (i) {
+      s.append(" ");
+    }
+    s.append(printValKind(wasm_valtype_kind(types->data[i])));
+  }
+  return s;
+}
+
+bool Wamr::link(std::string_view debug_name) {
+  assert(module_ != nullptr);
+
+  WasmImporttypeVec import_types;
```

Comment:
WAMR currently tends to link *.WAMR* automatically. We have to remove all the manual linking code. L305 -> L371